cpp/mg_utility/mage.hpp

```diff
@@ -0,0 +1,2757 @@
+
+/* #region Graph (Id, Graph, Nodes, GraphRelationships, Relationships, Properties & Labels) */
+
+/// Wrapper for int64_t IDs to prevent dangerous implicit conversions.
+class Id {
+ public:
+  Id() = default;
+
+  /// Construct Id from uint64_t
+  static Id FromUint(uint64_t id) { return Id(util::MemcpyCast<int64_t>(id)); }
+  /// Construct Id from int64_t
+  static Id FromInt(int64_t id) { return Id(id); }
+
+  int64_t AsInt() const { return id_; }
+  uint64_t AsUint() const { return util::MemcpyCast<uint64_t>(id_); }
+
+  bool operator==(const Id &other) const { return id_ == other.id_; }
+  bool operator!=(const Id &other) const { return !(*this == other); }
+
+  bool operator<(const Id &other) const { return id_ < other.id_; }
+
+ private:
+  explicit Id(int64_t id) : id_(id) {}
+
+  int64_t id_;
+};
+
+/// @brief Wrapper class for @ref mgp_graph.
+class Graph {
+ private:
+  friend class Node;
+  friend class Relationship;
+
+ public:
+  explicit Graph(mgp_graph *graph) : graph_(graph) {}
+
+  /// @brief Returns the graph order (number of nodes).
+  int64_t order() const;
+  /// @brief Returns the graph size (number of relationships).
+  int64_t size() const;
+
+  /// @brief Returns an iterable structure of the graph’s nodes.
+  GraphNodes nodes() const;
+  /// @brief Returns an iterable structure of the graph’s relationships.
+  GraphRelationships relationships() const;
+
+  /// @brief Returns the graph node with the given ID.
+  Node GetNodeById(const Id node_id) const;
+
+  /// @brief Returns whether the graph contains a node with the given ID.
+  bool Contains(const Id node_id) const;
+  /// @brief Returns whether the graph contains the given node.
+  bool Contains(const Node &node) const;
+  /// @brief Returns whether the graph contains the given relationship.
+  bool Contains(const Relationship &relationship) const;
+
+  /// @brief Returns whether the graph is mutable.
+  bool IsMutable() const { return mgp::graph_is_mutable(graph_); }
+  /// @brief Creates a node and adds it to the graph.
+  void CreateNode() { mgp::graph_create_vertex(graph_, memory); }
+  /// @brief Deletes a node from the graph.
+  void DeleteNode(const Node &node);
+  /// @brief Deletes a node and all its incident edges from the graph.
+  void DetachDeleteNode(const Node &node);
+  /// @brief Creates a relationship of type `type` between nodes `from` and `to` and adds it to the graph.
+  void CreateRelationship(const Node &from, const Node &to, const std::string_view type);
+  /// @brief Deletes a relationship from the graph.
+  void DeleteRelationship(const Relationship &relationship);
+
+ private:
+  mgp_graph *graph_;
+};
+
+/// @brief View of graph nodes; wrapper class for @ref mgp_vertices_iterator.
+class Nodes {
+ public:
+  explicit Nodes(mgp_vertices_iterator *nodes_iterator) : nodes_iterator_(nodes_iterator) {}
+
+  class Iterator {
+   public:
+    friend class Nodes;
+
+    explicit Iterator(mgp_vertices_iterator *nodes_iterator);
+    ~Iterator();
```

Comment:
Do you need a destructor? If you are releasing memory, then yes I presume, but then also define copy ctor and move ctor and copy assignment operator and move assignment operator

Reply:
Done